atc/policy/checker.go

```diff
@@ -0,0 +1,224 @@
+		f.ActionWhiteList = strings.Split(f.ActionWhiteList[0], ",")
+	}
+
+	if len(f.ActionBlackList) == 1 {
+		f.ActionBlackList = strings.Split(f.ActionBlackList[0], ",")
+	}
+
+	return f
+}
+
+type PolicyCheckInput struct {
+	Service        string      `json:"service"`
+	ClusterName    string      `json:"cluster_name"`
+	ClusterVersion string      `json:"cluster_version"`
+	HttpMethod     string      `json:"http_method,omitempty"`
+	Action         string      `json:"action"`
+	User           string      `json:"user"`
+	Team           string      `json:"team,omitempty"`
+	Pipeline       string      `json:"pipeline,omitempty"`
+	Data           interface{} `json:"data,omitempty"`
+}
+
+// PreChecker runs filters first, then calls underlying policy agent.
+type PreChecker interface {
+	// CheckHttpApi returns true if passes policy check. If not goes through
+	// policy check, just return true.
+	CheckHttpApi(string, string, *http.Request) (bool, error)
+
+	CheckTask(db.Build, atc.TaskConfig) (bool, error)
+}
+
+// Checker should be implemented by policy agents.
+type Checker interface {
+	// Check returns true if passes policy check. If not goes through policy
+	// check, just return true.
+	Check(PolicyCheckInput) (bool, error)
+}
+
+type CheckerFactory interface {
+	Description() string
+	IsConfigured() bool
+	NewChecker() (Checker, error)
+}
+
+var checkerFactories []CheckerFactory
+
+func RegisterChecker(factory CheckerFactory) {
+	checkerFactories = append(checkerFactories, factory)
+}
+
+func WireCheckers(group *flags.Group) {
+	for _, factory := range checkerFactories {
+		_, err := group.AddGroup(fmt.Sprintf("Policy Checker (%s)", factory.Description()), "", factory)
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+var (
+	clusterName    string
+	clusterVersion string
+)
+
+func Initialize(logger lager.Logger, cluster string, version string, filter Filter) (PreChecker, error) {
+	logger.Debug("policy-checker-initialize")
+
+	clusterName = cluster
+	clusterVersion = version
+
+	var checkerDescriptions []string
+	for _, factory := range checkerFactories {
+		if factory.IsConfigured() {
+			checkerDescriptions = append(checkerDescriptions, factory.Description())
+		}
+	}
+	if len(checkerDescriptions) > 1 {
+		return nil, fmt.Errorf("Multiple emitters configured: %s", strings.Join(checkerDescriptions, ", "))
+	}
+
+	for _, factory := range checkerFactories {
+		if factory.IsConfigured() {
+			realChecker, err := factory.NewChecker()
+			if err != nil {
+				return nil, err
+			}
+			return &preChecker{
+				filter:      filter.normalize(),
+				realChecker: realChecker,
+			}, nil
+		}
+	}
+
+	// No policy checker configured.
+	return nil, nil
+}
+
+type preChecker struct {
+	filter      Filter
+	realChecker Checker
+}
+
+func (c *preChecker) CheckHttpApi(action string, user string, req *http.Request) (bool, error) {
+	if c == nil {
+		return true, nil
+	}
+
+	// Ignore self invoked API calls.
+	if user == "system" {
```

Comment:
https://github.com/concourse/concourse/pull/4950 changes the definition of what we consider to be a "system" request. 

These values are now configurable:
https://github.com/concourse/concourse/blob/0a2bff940f795bccb0d7c02d6c63e6f90456220f/atc/api/accessor/accessor.go#L59-L60

So we may have to do something similar here.